src/lumigo_tracer/lambda_tracer/lambda_reporter.py

```diff
@@ -163,33 +166,62 @@ def report_json(
             if not edge_connection:
                 get_logger().warning("Can not establish connection. Skip sending span.")
                 return duration
+
+    lock = threading.Lock()
+
+    def send_single_span(data: str, retry: bool = True) -> None:
+        """
+        Helper function to send a single span request and handle retries,
+        including re-establishing connection if necessary.
+        """
+        global edge_connection
+        if edge_connection is None:
+            raise ValueError("Connection is not established")
+
+        try:
+            get_logger().debug(f"Sending data to {host}/{EDGE_PATH}: {data}")
+            edge_connection.request(
+                "POST",
+                EDGE_PATH,
+                data,
+                headers={
+                    "Content-Type": "application/json",
+                    "Authorization": Configuration.token or "",
+                },
+            )
+            response = edge_connection.getresponse()
+            response.read()  # We must read the response to keep the connection available
+            get_logger().info(f"Successful reporting, code: {getattr(response, 'code', 'unknown')}")
+        except socket.timeout:
+            get_logger().exception(f"Timeout while connecting to {host}")
+            InternalState.mark_timeout_to_edge()
+            internal_analytics_message("report: socket.timeout")
+        except Exception as e:
+            if retry:
+                get_logger().info(f"Could not report to {host}: ({str(e)}). Retrying.")
+                # time.sleep(1)  # TODO Introduce a backoff delay before retrying
+                with lock:
+                    edge_connection = establish_connection(host)  # Re-establish connection safely
+                send_single_span(data, retry=False)
+            else:
+                get_logger().exception("Could not report: A span was lost.", exc_info=e)
+                internal_analytics_message(f"report: {type(e)}")
+
     try:
         start_time = time.time()
-        edge_connection.request(
-            "POST",
-            EDGE_PATH,
-            to_send,
-            headers={
-                "Content-Type": "application/json",
-                "Authorization": Configuration.token or "",
-            },
-        )
-        response = edge_connection.getresponse()
-        response.read()  # We must read the response to keep the connection available
+
+        to_send_list: List[str] = to_send if isinstance(to_send, list) else [to_send]
+        # Send all the spans concurrently using a thread pool
+        # TODO Limit the number of threads, e.g., to N max_workers=N
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            futures = [executor.submit(send_single_span, span_data) for span_data in to_send_list]
+            # Wait for all the futures to complete
+            concurrent.futures.wait(futures)
+
```

Comment:
Don't run in threads if there is only 1 request to send...

Reply:
Changed to use a loop

Reply:
Changed to a loop